### ISSUE TITLE
Examples cpp11/container.cpp are bugged

### ISSUE THREAD

**Issue by ropieur** [msgpack/msgpack-c#806]
As stringstream::str() returns a string copy of the internal representation, data() returns a handle on a temporary object potentially destroyed inside unpack call. Actually the double calls to ss.str() refers two different objects.
    msgpack::object_handle oh = msgpack::unpack(ss.str().data(), ss.str().size());

**Comment by redboltz** [comment]
Thank you reporting the issue.

> As stringstream::str() returns a string copy of the internal representation, data() returns a handle on a temporary object potentially destroyed inside unpack call.

I don't think so. See https://stackoverflow.com/questions/584824/guaranteed-lifetime-of-temporary-in-c
Demo: https://wandbox.org/permlink/FsYof21iSoYzEkOb

Am I missing something?

> Actually the double calls to ss.str() refers two different objects.


True, but even if they are the different objects, the value of `size()` is the same, and one objects pointer is used. It works without any problems.

However, this is not a good style of programming.

So I will fix the code from

```cpp
msgpack::object_handle oh = msgpack::unpack(ss.str().data(), ss.str().size());
```

to 

```cpp
std::string const& str = ss.str();
msgpack::object_handle oh = msgpack::unpack(str.data(), str.size());
```

**Comment by ropieur** [comment]
A compiler is not forced to keep the temporaries. And that happened to me with my g++ compiler. The temporaries were destroyed once inside unpack. With the 2nd writing, this will indeed fix the issue because you force the compiler to maintain the lifetime of the temporary as long as the one of the reference.

**Comment by ropieur** [comment]
Well seems that https://stackoverflow.com/questions/584824/guaranteed-lifetime-of-temporary-in-c contradicts what I just stated. I played with your example and I couldn't reproduce the issue. So I must admit that I should be wrong and to review a bit more the C++ standard.
I got the issue in a context of streaming operations (operator<<) where I did call exactly the same way (ss.str().data(), ss.str().size()), but which is indeed different from a simple function call (msgpack::unpack)
